File: packages/totem-sdk/packages/governance/src/snapshot.ts
import type { MembershipEntry, MembershipSnapshot } from './types.js'
import { computeSnapshotHash } from './ids.js'

export function freezeMembershipSnapshot(
  daoId: string,
  entries: MembershipEntry[],
  frozenAt?: number,
): MembershipSnapshot {
  const at = frozenAt ?? Date.now()
  const map = new Map<string, MembershipEntry>()
  for (const entry of entries) {
    if (!entry.memberId) continue
    const existing = map.get(entry.memberId)
    if (!existing || entry.addedAt >= existing.addedAt) {
      map.set(entry.memberId, { ...entry })
    }
  }
  const snapshotEntries = Array.from(map.values())
  const hash = computeSnapshotHash(
    daoId,
    at,
    snapshotEntries.map((e) => ({ memberId: e.memberId, weight: e.weight })),
  )
  return { daoId, frozenAt: at, entries: map, hash }
}

export function verifyMembershipSnapshot(snapshot: MembershipSnapshot): boolean {
  const snapshotEntryHashes = Array.from(snapshot.entries.values()).map((e) => ({
    memberId: e.memberId,
    weight: e.weight,
  }))
  const expectedHash = computeSnapshotHash(snapshot.daoId, snapshot.frozenAt, snapshotEntryHashes)
  return expectedHash === snapshot.hash
}

export function getMemberWeight(
  snapshot: MembershipSnapshot,
  memberId: string,
): number {
  const entry = snapshot.entries.get(memberId)
  if (!entry) return 0
  if (entry.expiresAt !== undefined && Date.now() > entry.expiresAt) return 0
  return entry.weight
}

export function getTotalWeight(snapshot: MembershipSnapshot): number {
  let total = 0
  for (const entry of snapshot.entries.values()) {
    if (entry.expiresAt !== undefined && Date.now() > entry.expiresAt) continue
    total += entry.weight
  }
  return total
}

File: packages/totem-sdk/packages/governance/rust/src/snapshot.rs
use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::ids::compute_snapshot_hash;
use crate::types::{MembershipEntry, MembershipSnapshot};

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

pub fn freeze_membership_snapshot(
    dao_id: &str,
    entries: &[MembershipEntry],
    frozen_at: Option<u64>,
) -> MembershipSnapshot {
    let at = frozen_at.unwrap_or_else(now_ms);
    let mut map: HashMap<String, MembershipEntry> = HashMap::new();

    for entry in entries {
        if entry.member_id.is_empty() {
            continue;
        }
        if let Some(existing) = map.get(&entry.member_id) {
            if entry.added_at < existing.added_at {
                continue;
            }
        }
        map.insert(entry.member_id.clone(), entry.clone());
    }

    let snapshot_entries: Vec<(&str, f64)> = map
        .iter()
        .map(|(id, e)| (id.as_str(), e.weight))
        .collect();

    let hash = compute_snapshot_hash(dao_id, at, &snapshot_entries);

    MembershipSnapshot {
        dao_id: dao_id.to_string(),
        frozen_at: at,
        entries: map,
        hash,
    }
}

pub fn verify_membership_snapshot(snapshot: &MembershipSnapshot) -> bool {
    let entries: Vec<(&str, f64)> = snapshot
        .entries
        .iter()
        .map(|(id, e)| (id.as_str(), e.weight))
        .collect();
    let expected_hash = compute_snapshot_hash(&snapshot.dao_id, snapshot.frozen_at, &entries);
    expected_hash == snapshot.hash
}

pub fn get_member_weight(snapshot: &MembershipSnapshot, member_id: &str) -> f64 {
    match snapshot.entries.get(member_id) {
        Some(entry) => {
            if let Some(expires_at) = entry.expires_at {
                if now_ms() > expires_at {
                    return 0.0;
                }
            }
            entry.weight
        }
        None => 0.0,
    }
}

pub fn get_total_weight(snapshot: &MembershipSnapshot) -> f64 {
    let now = now_ms();
    snapshot
        .entries
        .values()
        .filter(|entry| {
            if let Some(expires_at) = entry.expires_at {
                now <= expires_at
            } else {
                true
            }
        })
        .map(|entry| entry.weight)
        .sum()
}

pub fn get_member_weight_at(
    snapshot: &MembershipSnapshot,
    member_id: &str,
    at: u64,
) -> f64 {
    match snapshot.entries.get(member_id) {
        Some(entry) => {
            if let Some(expires_at) = entry.expires_at {
                if at > expires_at {
                    return 0.0;
                }
            }
            entry.weight
        }
        None => 0.0,
    }
}

pub fn get_total_weight_at(snapshot: &MembershipSnapshot, at: u64) -> f64 {
    snapshot
        .entries
        .values()
        .filter(|entry| {
            if let Some(expires_at) = entry.expires_at {
                at <= expires_at
            } else {
                true
            }
        })
        .map(|entry| entry.weight)
        .sum()
}
